src/librustdoc/test.rs

```diff
@@ -432,11 +432,38 @@ impl Collector {
         }
     }
 
-    pub fn add_old_test(&mut self, line: usize, filename: String) {
-        let name = self.generate_name(line, &filename);
-        self.old_tests.push(name);
+    // to be removed once hoedown is gone
+    fn generate_name_beginning(&self, filename: &str) -> String {
+        if self.use_headers {
+            if let Some(ref header) = self.current_header {
+                format!("{} - {} (line", filename, header)
+            } else {
+                format!("{} - (line", filename)
+            }
+        } else {
+            format!("{} - {} (line", filename, self.names.join("::"))
+        }
     }
 
+pub fn add_old_test(&mut self, mut line: usize, filename: String) {
+    // hack to make it work in hoedown because the correct line gets lost when too many
+    // doc tests are present in one doc block.
+    let tmp_name = self.generate_name_beginning(&filename);
+    let mut number = 0;
+    if line > 1 {
+        for test in &self.old_tests {
+            if test.starts_with(&tmp_name) {
+                number += 1;
+            }
+        }
+    }
+    if number < line {
+        line -= number;
+    }
```

Comment:
What cases were you running into where the number of tests was more than the line number itself? Or is this a guard to prevent underflowing a usize?